patches/net/minecraft/server/network/ServerCommonPacketListenerImpl.java.patch

```diff
@@ -1,14 +1,80 @@
++        this.send(new net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket(packetPayload));
++    }
  
-     @Override
-     public void handleCustomPayload(ServerboundCustomPayloadPacket p_294276_) {
-+        if (p_294276_.payload() instanceof net.neoforged.neoforge.network.custom.payload.SimplePayload simplePayload && net.neoforged.neoforge.network.NetworkHooks.onCustomPayload(p_294276_, simplePayload, connection)) {
++    @Override
+     public void send(Packet<?> p_295099_, @Nullable PacketSendListener p_296321_) {
++        if (!net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().canSendPacket(p_295099_, this)) {
 +            return;
 +        }
 +        
-+        LOGGER.warn("Unknown custom packet payload: {}", p_294276_.payload().id());
+         boolean flag = !this.suspendFlushingOnServerThread || !this.server.isSameThread();
+ 
+         try {
+@@ -121,6 +_,11 @@
+             throw new ReportedException(crashreport);
+         }
      }
++    
++    @Override
++    public void send(net.minecraft.network.protocol.common.custom.CustomPacketPayload packetPayload, @Nullable PacketSendListener listener) {
++        this.send(new net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket(packetPayload), listener);
++    }
  
-     @Override
+     public void disconnect(Component p_294116_) {
+         this.connection.send(new ClientboundDisconnectPacket(p_294116_), PacketSendListener.thenRun(() -> this.connection.disconnect(p_294116_)));
+@@ -143,7 +_,41 @@
+         return this.latency;
+     }
+ 
++    /**
++     * Creates a new cookie for this connection.
++     *
++     * @param p_301973_ The client information.
++     * @return The cookie.
++     * @deprecated Use {@link #createCookie(ClientInformation, boolean)} instead, keeping the connection type information available.
++     */
++    @Deprecated
+     protected CommonListenerCookie createCookie(ClientInformation p_301973_) {
+-        return new CommonListenerCookie(this.playerProfile(), this.latency, p_301973_);
++        return new CommonListenerCookie(this.playerProfile(), this.latency, p_301973_, false);
++    }
++    
++    /**
++     * Creates a new cookie for this connection.
++     *
++     * @param p_301973_ The client information.
++     * @param isModdedConnection Whether the connection is modded.
++     * @return The cookie.
++     */
++    protected CommonListenerCookie createCookie(ClientInformation p_301973_, boolean isModdedConnection) {
++        return new CommonListenerCookie(this.playerProfile(), this.latency, p_301973_, isModdedConnection);
++    }
++    
++    @Override
++    public Connection getConnection() {
++        return connection;
++    }
++
++    @Override
++    public net.minecraft.util.thread.ReentrantBlockableEventLoop<?> getMainThreadEventLoop() {
++        return server;
++    }
++
++    @Override
++    public ServerCommonPacketListener self() {
```

Comment:
Why is this done here rather than how we do other extension interfaces of having the `self` method be private and just cast to the expected type?

Reply:
See other